### PR TITLE
[typehints] tests: enable mypy for linkcheck builder tests.

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Refactoring

### Purpose
- Improve coverage of typechecking through the codebase.

### Detail
- Enable `mypy` typechecking for the `tests/test_builders/test_build_linkcheck.py` file.

### Relates
- Indirectly supports #12126.

### CODE REVIEW

File: tests/test_builders/test_build_linkcheck.py
Comment (danieleades) [inline tests/test_builders/test_build_linkcheck.py:157]: would a `TypedDict` or even a dataclass be better here?

File: tests/test_builders/test_build_linkcheck.py
Comment (danieleades) [inline tests/test_builders/test_build_linkcheck.py:112]: i'm personally for this change, but it's worth noting that mypy supports pytest fixtures, and you can get IDE support for the types of test arg even without annotations, so these are not strictly required.

That said, i'm in favour of them anyway

File: tests/test_builders/test_build_linkcheck.py
Comment (danieleades) [inline tests/test_builders/test_build_linkcheck.py:709]: you *could* apply ruff/black-style formatting here as you go

File: tests/test_builders/test_build_linkcheck.py
Comment (danieleades) [inline tests/test_builders/test_build_linkcheck.py]: why have you used `confoverrides` above but not here?

**Review by danieleades — APPROVED** [review]

**Review by danieleades — CHANGES_REQUESTED** [review]
two more comments:

1. should this wait until after #12012 is merged?
2. can you modify the mypy config so that these files are not whitelisted?

**Comment by jayaddison** [comment]
Thanks @danieleades for the review!  I'll respond to the other questions soon; for now:

> two more comments:
> 
>     1. should this wait until after [use 'types-docutils' instead of 'docutils-stubs' for docutils type annotations #12012](https://github.com/sphinx-doc/sphinx/pull/12012) is merged?

Yep, I'm OK to wait for that :+1: 
 
>     2. can you modify the mypy config so that these files are not whitelisted?

Odd, I thought I had done that.  But I will push that change, thanks!

**Comment by jayaddison** [comment]
> > 2. can you modify the mypy config so that these files are not whitelisted?
> > ```
> 
> Odd, I thought I had done that. But I will push that change, thanks!

Ah.  I _did_ do that in commit 56ef23b0672dc38def3e2ad43e5355341dcc3bf7 - but then fumbled it by accidentally including a removal of that change in 5fed4b3f986641480550cd455af5c3f5ddb289a7.

Re-applying 56ef23b0672dc38def3e2ad43e5355341dcc3bf7 ...

File: pyproject.toml
Comment (danieleades) [inline pyproject.toml]: i think this can be removed rather than just commented out

**Review by danieleades — CHANGES_REQUESTED** [review]

File: tests/test_builders/test_build_linkcheck.py
Comment (jayaddison) [reply tests/test_builders/test_build_linkcheck.py:157]: It would do, yep; this is not particularly robust type information.  I think I'd prefer to get some type-hints, even if non-ideal, in place and to use those to assist refactoring.

File: tests/test_builders/test_build_linkcheck.py
Comment (jayaddison) [reply tests/test_builders/test_build_linkcheck.py:112]: Thanks - I find them fairly useful when reading the source 'as-is' in plaintext editors (some of the more complicated type-hints can become distracting or take up screen space, but generally I find they're possible to narrow down to be relatively concise).

File: tests/test_builders/test_build_linkcheck.py
Comment (jayaddison) [reply tests/test_builders/test_build_linkcheck.py:709]: I'll keep that on the todo-list; I have #12126 open that changes a bunch of the `test_build_linkcheck` code, so I'd prefer not to introduce many-line conflicts (a few limited/construtive edit conflicts are OK, but reformatting looks like it'd be larger).

File: tests/test_builders/test_build_linkcheck.py
Comment (jayaddison) [reply tests/test_builders/test_build_linkcheck.py]: The truth is that I've read and re-read most of the tests in this file, but these few here -- the ones that don't involve test HTTP servers -- remain a bit of a mystery to me.  That means I'm less inclined to make multi-line or multi-word edits; I'm trying to use the smallest change possible because I don't feel like a maintainer of these.

File: pyproject.toml
Comment (jayaddison) [reply pyproject.toml]: Taking that discussion to: https://github.com/sphinx-doc/sphinx/pull/12109#discussion_r1534412739 (also has a circular comment-reference back to here)

File: tests/test_builders/test_build_linkcheck.py
Comment (picnixz) [reply tests/test_builders/test_build_linkcheck.py:112]: Personally, I think having a coverage on the test function themselves is a bit too much (considering the number of test functions to add).

As such, I think it is preferrable, in a test file, to only type when you are using a non-test function (utility, fixture, etc, but not a `test_*`).

Now, mypy does **not** know about the type (PyCharm does know about it because I have autocompletion but `reveal_type` gives me `Any`) so it could be fine; however, your IDE could technically help you out for this.

File: pyproject.toml
Comment (picnixz) [reply pyproject.toml]: After experimenting, removing or commenting out induce merge conflicts either way. So I'm for removing them (less code to review, and files would anyway be printed out when you mypy on tests).

File: tests/test_builders/test_build_linkcheck.py
Comment (picnixz) [reply tests/test_builders/test_build_linkcheck.py]: Those tests are actually using the `html` builder (but don't build anything since you don't call 'build()') and the `test-root` but actually, we are using an `app` because we just want to get its default configuration... 

I feel that we need a fixture with a simple configuration instead.

File: tests/test_builders/test_build_linkcheck.py
Comment (picnixz) [inline tests/test_builders/test_build_linkcheck.py]: Funny that you need this kind of annotation.

File: tests/test_builders/test_build_linkcheck.py
Comment (picnixz) [reply tests/test_builders/test_build_linkcheck.py:157]: Neither of them. It's better to have an in-place value since it doesn't appear to be used more than once.

File: tests/test_builders/test_build_linkcheck.py
Comment (jayaddison) [reply tests/test_builders/test_build_linkcheck.py]: Ok; recently I was reconsidering the choice _not_ to move this into `confoverrides`  - and it seems like there's more consensus that it would make sense to use `confoverrides`, so I'll push that change here soon.

File: pyproject.toml
Comment (jayaddison) [reply pyproject.toml]: This was a misunderstanding on my behalf; I had noticed the existing commented-out entries, but misinterpreted those as file/module names, and thought that they indicated modules that had already been converted to add type-hinting.

In short: I'll remove this line.

File: tests/test_builders/test_build_linkcheck.py
Comment (picnixz) [reply tests/test_builders/test_build_linkcheck.py]: Oh no, for this one it's fine. What I meant is that we should not need to have a `app` to get a configuration but be able to have a standalone `config` object.

File: tests/test_builders/test_build_linkcheck.py
Comment (jayaddison) [reply tests/test_builders/test_build_linkcheck.py:157]: I'm considering re-using that helper method elsewhere - but it shouldn't be included in this changeset (1. too large a refactor, 2. I haven't explored it to know what's possible yet, 3. it would introduce conflicts elsewhere).

Comment (picnixz) [reply tests/test_builders/test_build_linkcheck.py:157]: If you want to reuse it, put it in an util file (and not in the test file itself if you know that it will be imported)

File: tests/test_builders/test_build_linkcheck.py
Comment (jayaddison) [reply tests/test_builders/test_build_linkcheck.py:112]: I'm sympathetic to all the (no pun intended) arguments.  Less typing seems good; return types shouldn't matter on `test_` methods because they shouldn't be called from other code -- on the other hand, visually and for some utilities the hints could be useful, and they could catch unintended errors (although arguably some of those will be noise).

On the principle of preferring fewer special-cases for code style, I think I'll keep the type hints on the `test_` functions.

**Comment by jayaddison** [comment]
> two more comments:
> 
>     1. should this wait until after [use 'types-docutils' instead of 'docutils-stubs' for docutils type annotations #12012](https://github.com/sphinx-doc/sphinx/pull/12012) is merged?

With that merged, I'm going to merge mainline again into this branch (`mypy` linting looks good after a fresh install of the lint+test dependencies here).
 
>     2. can you modify the mypy config so that these files are not whitelisted?

Also resolved - thanks again!

File: tests/test_builders/test_build_linkcheck.py
Comment (picnixz) [reply tests/test_builders/test_build_linkcheck.py:112]: I think, typing the tests (since mypy doesn't know about them) could effectively be a good idea. Or, alternatively, I would suggest having a mypy plugin for that because it would really be easier (where you would complete the signatures yourself since you *know* exactly that 1) the test functions are all prefixed by `test_` and are given in a `test_*.py` file 2) the known fixtures are named the same and their type is *known*).

File: tests/test_builders/test_build_linkcheck.py
Comment (picnixz) [inline tests/test_builders/test_build_linkcheck.py]: This should be a `@staticmethod` or something outside of this class.

File: tests/test_builders/test_build_linkcheck.py
Comment (picnixz) [inline tests/test_builders/test_build_linkcheck.py:977]: If you want to keep it inline:

```suggestion
        begin, checked = time.time(), []  # type: (float, list[CheckResult])
```

Comment (jayaddison) [reply tests/test_builders/test_build_linkcheck.py:977]: Nice to know that kind of typehint is possible; I think one-per-line is probably simpler here though.  And that `time` stuff could be worth revisiting later, so separating the line has some benefit.

Comment (picnixz) [reply tests/test_builders/test_build_linkcheck.py:977]: It's a legacy way to do it (the 2.7 way to do when `x: type` was not available). Note that there is an issue with this construction with mypy when you are using constrained generics (it's an issue I discovered as well and patched but it's not merged yet: https://github.com/python/mypy/issues/16649).

**Comment by picnixz** [comment]
Could I ask you some insights on https://github.com/sphinx-doc/sphinx/pull/12198? because depending on the direction, we could just have a smaller diff (at the cost of the mypy plugin).

In addition, I'm considering merging https://github.com/sphinx-doc/sphinx/pull/12199 today as well (see the rationale in the PR).

**Comment by jayaddison** [comment]
> ~Could I ask you some insights on #12198? because depending on the direction, we could just have a smaller diff (at the cost of the mypy plugin).~ Mypy plugins do not give me the possibility to annotate an un-annotated function _after_ :(
> 
> In addition, I'm considering merging #12199 today as well (see the rationale in the PR).

Ok, thanks @picnixz.  Since this affects other branches, my preferred order to merge would be:

1. #12199 (adjust the config exclude list style)
2. #12160 (this PR - enable typehinting for the `linkcheck` tests)
3. #12126 (resolve some conflicts, and then land the port-isolation with type-hinting also)

(and oops, I missed your edit before commenting on that PR - but I could help out with attempting the copyright method refactor if that'd be useful)

**Comment by danieleades** [comment]
there's a significant downside to this approach compared to the `[[tool.mypy.overrides]]` approach, which is that for the `[[tool.mypy.overrides]]` approach you can shrink the whitelist not just be removing files from the whitelist, but also by making the ignored errors more fine-grained (for example by changing it from ignoring all errors to only ignoring some classes of errors)

**Comment by picnixz** [comment]
>  (for example by changing it from ignoring all errors to only ignoring some classes of errors)

For this one, I would suggest (since we are in a test-file) using a top-file comment.

**Review by chrisjsewell — APPROVED** [review]
All good cheers

**Pull request merged**
